### PR TITLE
[BEAM-430] Add GcpTempLocation, and remove defaulting tempLocation to stagingLocation in DataflowRunner

### PR DESCRIPTION
**Pull request by peihe**

### CODE REVIEW

**Comment by peihe** [comment]
R: @lukecwik for tempLocation and pipeline options related code
R: @dhalperi for tempLocation and BigQuery related code

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineRunner.java [hunk 1]
```diff
@@ -260,26 +260,11 @@ public static DataflowPipelineRunner fromOptions(PipelineOptions options) {
...
-    } else if (Strings.isNullOrEmpty(dataflowOptions.getStagingLocation())) {
-      try {
-        dataflowOptions.setStagingLocation(
-            IOChannelUtils.resolve(dataflowOptions.getTempLocation(), "staging"));
-      } catch (IOException e) {
-        throw new IllegalArgumentException("Unable to resolve PipelineOptions.stagingLocation "
-            + "from PipelineOptions.tempLocation. Please set the staging location explicitly.", e);
-      }
+    if (dataflowOptions.getStagingLocation() != null) {
+      validator.validateOutputFilePrefixSupported(dataflowOptions.getStagingLocation());
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineRunner.java]: Don't you still want the code that sets `stagingLocation` from `tempLocation`?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java [hunk 2]
```diff
@@ -29,14 +29,19 @@
 import org.apache.beam.sdk.options.PubsubOptions;
 import org.apache.beam.sdk.options.StreamingOptions;
 import org.apache.beam.sdk.options.Validation;
+import org.apache.beam.sdk.util.IOChannelUtils;
 
 import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: please use static imports

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java [hunk 3]
```diff
@@ -69,6 +74,7 @@
       + "Must be a valid Cloud Storage URL, beginning with the prefix \"gs://\". "
       + "At least one of stagingLocation or tempLocation must be set. If stagingLocation is unset, "
       + "defaults to using tempLocation.")
+  @Default.InstanceFactory(StagingLocationFactory.class)
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: why not default to `gcpTempLocation` instead of `tempLocation`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java [hunk 4]
```diff
@@ -550,13 +550,13 @@ private static void dryRunQuery(BigQueryOptions options, String query) {
         final BigQueryServices bqServices = getBigQueryServices();
 
         final String extractDestinationDir;
-        String tempLocation = bqOptions.getTempLocation();
+        String bqTempLocation = bqOptions.getBigQueryTempLocation();
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java]: I think that this is not necessary. Would rather just default to `GcpTempLocation` and let users additionally override the temp location for each individual BQ read or write.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java [hunk 5]
```diff
@@ -77,9 +77,11 @@
   public static TestDataflowRunner fromOptions(
       PipelineOptions options) {
     TestDataflowPipelineOptions dataflowOptions = options.as(TestDataflowPipelineOptions.class);
-    dataflowOptions.setStagingLocation(Joiner.on("/").join(
+    String tempLocation = Joiner.on("/").join(
         new String[]{dataflowOptions.getTempRoot(),
-          dataflowOptions.getJobName(), "output", "results"}));
+        dataflowOptions.getJobName(), "output", "results"});
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: line wrap?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java [hunk 6]
```diff
@@ -77,9 +77,11 @@
   public static TestDataflowRunner fromOptions(
       PipelineOptions options) {
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: line wrap?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java [hunk 7]
```diff
@@ -77,9 +77,11 @@
...
     TestDataflowPipelineOptions dataflowOptions = options.as(TestDataflowPipelineOptions.class);
-    dataflowOptions.setStagingLocation(Joiner.on("/").join(
+    String tempLocation = Joiner.on("/").join(
         new String[]{dataflowOptions.getTempRoot(),
-          dataflowOptions.getJobName(), "output", "results"}));
+        dataflowOptions.getJobName(), "output", "results"});
+    dataflowOptions.setTempLocation(tempLocation);
+    dataflowOptions.setStagingLocation(tempLocation);
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: This to me says that we have something wrong in our initialization. Do we simply need to fix the Jenkins config?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java [hunk 8]
```diff
@@ -66,14 +67,14 @@
    *
    * <p>Must be a valid Cloud Storage URL, beginning with the prefix "gs://"
    *
-   * <p>At least one of {@link PipelineOptions#getTempLocation()} or {@link #getStagingLocation()}
+   * <p>At least one of {@link GcpOptions#getGcpTempLocation()} or {@link #getStagingLocation()}
    * must be set. If {@link #getStagingLocation()} is not set, then the Dataflow
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: I think this is no longer correct, right?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java (hunk 5 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java (hunk 6 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java (hunk 4 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java (hunk 2 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineRunner.java (hunk 1 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineRunner.java]: I moved it to the DefaultValueFactory

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java (hunk 3 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: The code does that.
Updated the doc to match the code.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java (hunk 8 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: updated

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java (hunk 7 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: stagingLocation is already set in here.
+@jasonkuster for comments on setting tempRoot vs tempLocation/stagingLocation in Jenkins

**Comment by peihe** [comment]
I rebased for a unrelated test failure and conflicts.
PTAL

**Comment by dhalperi** [comment]
R: @jasonkuster please take a look?

Pei, let me know once Jason has given approval and I can merge.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 9]
```diff
@@ -260,26 +260,11 @@ public static DataflowRunner fromOptions(PipelineOptions options) {
...
-    } else if (isNullOrEmpty(dataflowOptions.getStagingLocation())) {
-      try {
-        dataflowOptions.setStagingLocation(
-            IOChannelUtils.resolve(dataflowOptions.getTempLocation(), "staging"));
-      } catch (IOException e) {
-        throw new IllegalArgumentException("Unable to resolve PipelineOptions.stagingLocation "
-            + "from PipelineOptions.tempLocation. Please set the staging location explicitly.", e);
-      }
+    if (dataflowOptions.getStagingLocation() != null) {
+      validator.validateOutputFilePrefixSupported(dataflowOptions.getStagingLocation());
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: should we be setting staging location from temp location?

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java [hunk 10]
```diff
@@ -90,4 +93,39 @@ public void testUtf8UserNameAndApplicationNameIsNormalized() {
...
+    assertEquals("gs://temp_location/staging", options.getStagingLocation());
+  }
+
+  @Test
+  public void testDefaultStagingLocationInvalid() {
+    DataflowPipelineOptions options = PipelineOptionsFactory.as(DataflowPipelineOptions.class);
+    options.setTempLocation("file://temp_location");
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("GCP temp location requires a valid 'gs://' path");
+    options.getStagingLocation();
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java]: Tests of `options.getGcsTempLocation`?

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowRunnerTest.java [hunk 11]
```diff
@@ -189,7 +189,8 @@ private static Dataflow buildMockDataflow(
     return mockDataflowClient;
   }
 
-  private GcsUtil buildMockGcsUtil(boolean bucketExists) throws IOException {
+  private GcsUtil buildMockGcsUtil(Boolean bucketExist, Boolean... bucketExists)
```
Comment (jasonkuster) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowRunnerTest.java]: What is the difference between bucketexist and bucketexists? If the variable names cannot be clarified please add a comment explaining.

**Comment by jasonkuster** [comment]
Generally looks good. One nit - is GcpTempLocation the right name? Shouldn't it be GcsTempLocation?

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowRunnerTest.java (hunk 11 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowRunnerTest.java]: added

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java (hunk 10 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java]: You are right. The code and test are wrong. 
GcpTempLocation should be optional and not throw.
Updated

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java (hunk 9 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: Staging location is defaulting to gcp temp location in StagingLocationFactory. And, gcp temp location is defaulting to temp location if it is a valid gs path.

**Comment by peihe** [comment]
GcpTempLocation means the temp location in Google Cloud Platform.

I like it slightly better, since it is more broader. And, we are using it in BigQueryIO.

@dhalperi comments?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java (hunk 7 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: I chatted with @jasonkuster offline.
Jenkins is not the place to set stagingLocation, since stagingLocation is a dataflow pipeline options.
tempLocation might be able to replace tempRoot(), but this is out of the scope of this PR.

**Comment by peihe** [comment]
PTAL

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java [hunk 12]
```diff
@@ -90,4 +93,40 @@ public void testUtf8UserNameAndApplicationNameIsNormalized() {
     options.setAppName("fəˈnɛtık əsoʊsiˈeıʃn");
     assertEquals("f00n0t0k00so0si0e00n-0i00nt00n000n0l0-1208190706", options.getJobName());
   }
+
+  @Test
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java]: please do much more comprehensive testing here, including of `gcpTempLocation` and all reasonable error cases.

Comment (peihe) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java]: Added some asserts.

There is another change that validates gcpTempLocation and set it for TempStoragePrefix in DataflowRunner. I want to do that in a separate PR.

gcpTempLocation is still optional in DataflowPipelineOptions.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowRunnerTest.java [hunk 13]
```diff
@@ -189,7 +189,14 @@ private static Dataflow buildMockDataflow(
     return mockDataflowClient;
   }
 
-  private GcsUtil buildMockGcsUtil(boolean bucketExists) throws IOException {
+  /**
+   * Build a mock {@link GcsUtil} with return values.
+   *
+   * @param bucketExist fist return value
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowRunnerTest.java]: first

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java [hunk 14]
```diff
@@ -74,12 +74,15 @@
...
-    dataflowOptions.setStagingLocation(Joiner.on("/").join(
-        new String[]{dataflowOptions.getTempRoot(),
-          dataflowOptions.getJobName(), "output", "results"}));
+    String tempLocation = Joiner.on("/").join(
+        dataflowOptions.getTempRoot(),
+        dataflowOptions.getJobName(),
+        "output",
+        "results");
+    dataflowOptions.setTempLocation(tempLocation);
+    dataflowOptions.setStagingLocation(tempLocation);
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: can we remove this line since defaults should work all the way down?

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java [hunk 15]
```diff
@@ -90,4 +94,42 @@ public void testUtf8UserNameAndApplicationNameIsNormalized() {
...
+  }
+
+  @Test
+  public void testDefaultStagingLocationUnset() {
+    DataflowPipelineOptions options = PipelineOptionsFactory.as(DataflowPipelineOptions.class);
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Missing required value: at least one of tempLocation, gcpTempLocation"
+        + " or stagingLocation must be set.");
+    options.getStagingLocation();
+  }
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java]: test where staging defaults to gcsTemp

e.g., temp is a gs://a, gcsTemp is gs://b, and staging picks gs://b/staging

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowRunnerTest.java (hunk 13 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowRunnerTest.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java (hunk 14 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java]: done

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java (hunk 15 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptionsTest.java]: done

**Comment by peihe** [comment]
PTAL

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 16]
```diff
@@ -260,27 +260,12 @@ public static DataflowRunner fromOptions(PipelineOptions options) {
     }
 
     PathValidator validator = dataflowOptions.getPathValidator();
-    checkArgument(!(isNullOrEmpty(dataflowOptions.getTempLocation())
-        && isNullOrEmpty(dataflowOptions.getStagingLocation())),
-        "Missing required value: at least one of tempLocation or stagingLocation must be set.");
-
+    if (dataflowOptions.getGcpTempLocation() != null) {
+      validator.validateOutputFilePrefixSupported(dataflowOptions.getGcpTempLocation());
+    }
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: If `gcpTempLocation` is still `null` here, the job will fail?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java [hunk 17]
```diff
@@ -61,14 +67,15 @@
    *
    * <p>Must be a valid Cloud Storage URL, beginning with the prefix "gs://"
    *
-   * <p>At least one of {@link PipelineOptions#getTempLocation()} or {@link #getStagingLocation()}
+   * <p>At least one of {@link GcpOptions#getGcpTempLocation()} or {@link #getStagingLocation()}
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: I think the real requirement now is that `gcpTempLocation` must be set, and `stagignLocation` can optionally also be set.

@francesperry FYI

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java [hunk 18]
```diff
@@ -61,14 +67,15 @@
...
    * must be set. If {@link #getStagingLocation()} is not set, then the Dataflow
-   * pipeline defaults to using {@link PipelineOptions#getTempLocation()}.
+   * pipeline defaults to using {@link GcpOptions#getGcpTempLocation()}.
    */
   @Description("GCS path for staging local files, e.g. \"gs://bucket/object\". "
       + "Must be a valid Cloud Storage URL, beginning with the prefix \"gs://\". "
-      + "At least one of stagingLocation or tempLocation must be set. If stagingLocation is unset, "
-      + "defaults to using tempLocation.")
+      + "At least one of stagingLocation or gcpTempLocation must be set. "
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: needs update

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java [hunk 19]
```diff
@@ -61,14 +67,15 @@
...
+      + "If stagingLocation is unset, defaults to using gcpTempLocation.")
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: defaults to gcpTempLocation/staging

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java [hunk 20]
```diff
@@ -123,4 +130,25 @@ public String create(PipelineOptions options) {
...
+   * Returns a default staging location under {@link GcpOptions#getGcpTempLocation}.
+   */
+  public static class StagingLocationFactory implements DefaultValueFactory<String> {
+
+    @Override
+    public String create(PipelineOptions options) {
+      String gcpTempLocation = options.as(GcpOptions.class).getGcpTempLocation();
+      checkArgument(!isNullOrEmpty(gcpTempLocation),
+          "Missing required value: at least one of tempLocation, gcpTempLocation "
+          + "or stagingLocation must be set.");
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: can be misleading, because `tempLocation` can be set to `file://blah` and this will still not work.

I would just say `Error constructing default value for stagingLocation: gcpTempLocation is missing. Either stagingLocation must be set explicitly or a valid value must be provided for gcpTempLocation`)

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java [hunk 21]
```diff
@@ -123,4 +130,25 @@ public String create(PipelineOptions options) {
...
+
+      try {
+        return IOChannelUtils.resolve(gcpTempLocation, "staging");
+      } catch (IOException e) {
+        throw new IllegalArgumentException("Unable to resolve stagingLocation from gcpTempLocation."
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: include the value of `gcpTempLocation` here.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java (hunk 16 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java (hunk 17 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java (hunk 18 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java (hunk 19 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java (hunk 20 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java (hunk 21 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: done

**Comment by peihe** [comment]
PTAL

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 22]
```diff
@@ -182,16 +182,17 @@
...
+ * A {@link PipelineRunner} that executes the operations in the pipeline by first translating them
+ * to the Dataflow representation using the {@link DataflowPipelineTranslator} and then submitting
  * them to a Dataflow service for execution.
  *
- * <p><h3>Permissions</h3>
- * When reading from a Dataflow source or writing to a Dataflow sink using
- * {@code DataflowRunner}, the Google cloudservices account and the Google compute engine
- * service account of the GCP project running the Dataflow Job will need access to the corresponding
- * source/sink.
+ * <p>
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: drop empty paragraph?

Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: merged into "Permissions"

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java [hunk 23]
```diff
@@ -61,14 +68,14 @@
...
-   * pipeline defaults to using {@link PipelineOptions#getTempLocation()}.
+   * <p>If {@link #getStagingLocation()} is not set, it will default to
+   * {@link GcpOptions#getGcpTempLocation()}. {@link GcpOptions#getGcpTempLocation()}
+   * must be a valid GCS path.
    */
   @Description("GCS path for staging local files, e.g. \"gs://bucket/object\". "
       + "Must be a valid Cloud Storage URL, beginning with the prefix \"gs://\". "
-      + "At least one of stagingLocation or tempLocation must be set. If stagingLocation is unset, "
-      + "defaults to using tempLocation.")
+      + "If stagingLocation is unset, defaults to gcpTempLocation.")
```
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: defaults to gcpTempLocation with or without /staging?

**Comment by dhalperi** [comment]
LGTM with 1 trivial tweak.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java (hunk 23 above)
Comment (peihe) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java]: updated
